lib/es6/Template.js

```diff
@@ -0,0 +1,251 @@
+  const { raw } = trimCommonWhitespaceFromLines(strings);
+
+  const delimiters = [];
+  const chunks = [];
+
+  const lexer = makeLexer();
+
+  let delimiter = null;
+  for (let i = 0, len = raw.length; i < len; ++i) {
+    let chunk = String(raw[i]);
+    if (delimiter === '`') {
+      // Treat raw \` in an identifier literal as an ending delimiter.
+      chunk = chunk.replace(/^([^\\`]|\\[\s\S])*\\`/, '$1`');
+    }
+    const newDelimiter = lexer(chunk);
+    if (newDelimiter === '`' && !delimiter) {
+      // Treat literal \` outside a string context as starting an
+      // identifier literal
+      chunk = chunk.replace(
+        /((?:^|[^\\])(?:\\\\)*)\\(`(?:[^`\\]|\\[\s\S])*)$/, '$1$2');
+    }
+
+    chunks.push(chunk);
+    delimiters.push(newDelimiter);
+    delimiter = newDelimiter;
+  }
+
+  if (delimiter) {
+    throw new Error(`Unclosed quoted string: ${delimiter}`);
+  }
+
+  return { raw, delimiters, chunks };
+}
+
+function interpolateSqlIntoFragment (
+  { raw, delimiters, chunks }, strings, values) {
+  // A buffer to accumulate output.
+  let [ result ] = chunks;
+  for (let i = 1, len = raw.length; i < len; ++i) {
+    const chunk = chunks[i];
+    // The count of values must be 1 less than the surrounding
+    // chunks of literal text.
+    if (i !== 0) {
+      const delimiter = delimiters[i - 1];
+      const value = values[i - 1];
+      if (delimiter) {
+        result += escapeDelimitedValue(value, delimiter);
+      } else {
+        result = appendValue(result, value, chunk);
+      }
+    }
+
+    result += chunk;
+  }
+
+  return new SqlFragment(result);
+}
+
+function escapeDelimitedValue (value, delimiter) {
+  if (delimiter === '`') {
+    return SqlString.escapeId(String(value)).replace(/^`|`$/g, '');
+  }
+  const escaped = SqlString.escape(String(value));
+  return escaped.substring(1, escaped.length - 1);
+}
+
+function appendValue (resultBefore, value, chunk) {
+  let needsSpace = false;
+  let result = resultBefore;
+  const valueArray = Array.isArray(value) ? value : [ value ];
+  for (let i = 0, nValues = valueArray.length; i < nValues; ++i) {
+    if (i) {
+      result += ', ';
+    }
+
+    const one = valueArray[i];
+    let valueStr = null;
+    if (one instanceof SqlFragment) {
+      if (!/(?:^|[\n\r\t ,\x28])$/.test(result)) {
+        result += ' ';
+      }
+      valueStr = one.toString();
+      needsSpace = i + 1 === nValues;
+    } else if (one instanceof Identifier) {
+      valueStr = SqlString.escapeId(one.toString());
+    } else {
+      // If we need to handle nested arrays, we would recurse here.
+      valueStr = SqlString.format('?', one);
+    }
+    result += valueStr;
+  }
+
+  if (needsSpace && chunk && !/^[\n\r\t ,\x29]/.test(chunk)) {
+    result += ' ';
+  }
+
+  return result;
+}
+
+/**
+ * Template tag function that contextually autoescapes values
+ * producing a SqlFragment.
+ */
+const sql = memoizedTagFunction(computeStatic, interpolateSqlIntoFragment);
```

Comment:
If this is an in memory cache, let's make sure to note in the README the maximum amount of RAM this is set to so folks will understand how much memory this will consume over the lifetime of their application.

Reply:
Underlying the cache is a `WeakMap`, so it should adapt to memory pressure.

The maximum memory usage should scale with the number of uses of the tag that appear in the source code.

Reply:
I'm not sure I understand, sorry. Doesn't memoizing something cache some aspect in memory for later use? What is the maximum number of objects that will end up in that cache?

Reply:
The maximum number of objects is 3 per occurrence of a relevant tag usage:

* Two created arrays
* One record that bundles the two together.

They're not pinned in memory since they key of the template object which is itself weakly referenced by the containing realm.

If that answers your question, I can put that in the docs.

Reply:
When <code>sql&#96;foo${bar}&#96;</code> is parsed by the EcmaScript engine, it creates a frozen array `['foo', '']`.

When control reaches that expression, the memoizer uses that array as a key in a `WeakMap` lookup to decide whether it has to run the lexer or whether it can reuse.

The frozen array is itself weakly held by the module scope.  This happens in ["12.2.9.3Runtime Semantics: GetTemplateObject ( templateLiteral )"](https://www.ecma-international.org/ecma-262/7.0/#sec-gettemplateobject).  Although the spec doesn't make this explicit, it has to be weakly held because otherwise

```js
function foo() {}
while (1) {
  eval('  foo`...`  ')
}
```
would leak memory.

So after a full garbage collection, any memoized state for scopes that are not reachable by any re-enterable scope will be flushed.

So the max size of the memo-table post full garbage collection should be the count of <code>sql&#96;...&#96;</code> calls in loaded modules.

By "re-enterable scope" above, I mean that those used only in `(function () { ... }())` module initializers and in `eval`-ed code should not contribute in the steady state.

The min size of the memo-table will be the count of those uses that have been evaluated.

The memory cost per entry in the memo table will be the 3 objects detailed earlier.

Reply:
I did a bit of digging around and found https://github.com/tc39/ecma262/issues/840 so some of my assumptions were misfounded.  I'm going to replace the cache with an LRU cache which should put a constant limit on the memory overhead.

Reply:
I replaced the WeakMap with an LRU cache.

Reply:
What is the LRU cache size in bytes set to / is it adjustable?

Reply:
The cache caps at 100 entries.

I'm going to assume that in the 3σ case, all uses fits on a 120 column line and have 12  `${...}`s.
In this 3σ case, a full cache consumes 11.7kB.

I'm going to assume that the average case has 3 or fewer `${...}`.
In the typical case, a full cache consumes 4.7kB.

If storing a single ASCII character in an array costs less than an 8B pointer, then it will be correspondingly cheaper.

I could probably save some space by joining the contexts into a string instead of storing characters in an array, but I don't have the tooling to do memory micro-benchmarks so can't say for sure.

The LRU cache size is not adjustable.  I could make that adjustable if you have an idea for an API.